### ISSUE TITLE
`lax.scan` is ~6x slower to run than hand-written loops

### ISSUE THREAD

**Issue by afrozenator** [google/jax#810]
This is on a GPU backend, haven't tried on others.

https://colab.research.google.com/drive/1N1jrvGNFRhTnYLOCL6vXUzJN3pRR3pdr

Minimal repro below
```
from jax import numpy as np
from jax import grad, jit, vmap, lax
from jax import random as jax_random

import numpy as onp

@jit
def rewards_to_go(rewards, mask, gamma=0.99):
  r"""Computes rewards to go.
  Args:
    rewards: np.ndarray of shape (B, T) of rewards.
    mask: np.ndarray of shape (B, T) of mask for the rewards.
    gamma: float, discount factor.

  Returns:
    rewards to go, np.ndarray of shape (B, T).
  """
  B, T = rewards.shape  # pylint: disable=invalid-name,unused-variable

  masked_rewards = rewards * mask  # (B, T)

  # Compute r2g_{T-1} at the start and then compute backwards in time.
  r2gs = [masked_rewards[:, -1]]

  # Go from T-2 down to 0.
  for t in reversed(range(T - 1)):
    r2gs.append(masked_rewards[:, t] + (gamma * r2gs[-1]))

  # The list should have length T.
  assert T == len(r2gs)

  # First we stack them in the correct way to make it (B, T), but these are
  # still from newest (T-1) to oldest (0), so then we flip it on time axis.
  return np.flip(np.stack(r2gs, axis=1), axis=1)


@jit
def scan_rewards_to_go(rewards, mask, gamma=0.99):
  masked_rewards = rewards * mask  # (B, T)

  reversed_rewards = np.flip(masked_rewards, axis=1)  # (B, T) flipped on time.
  rrt = np.transpose(reversed_rewards)  # (T, B) transpose to scan over time.

  def discounting_add(carry, reward):
    x = reward + (gamma * carry)
    return x, x

  _, ys = lax.scan(discounting_add,
                   np.zeros_like(rrt[0], dtype=np.float32),
                   rrt.astype(np.float32))

  # ys is (T, B) and T is in reverse order.
  return np.flip(np.transpose(ys), axis=1)

B, T = 16, 128

num_examples = 100
rewards = []
pvs = []
mask = []

for _ in range(num_examples):
  rewards.append(onp.random.randn(B, T))
  pvs.append(onp.random.randn(B, T+1))
  ones = onp.full((B, T), 1, dtype=onp.int32)
  for one in ones:
    l = onp.random.randint(0, T)
    one[range(l,T)] = 0
  mask.append(ones)
```

Now time the invocations:

```
%timeit [rewards_to_go(rewards[i], mask[i]) for i in range(num_examples)]
```

and 

```
%timeit [scan_rewards_to_go(rewards[i], mask[i]) for i in range(num_examples)]
```

**Comment by cshesse** [comment]
I have a related scan file here: https://gist.github.com/cshesse/73f27d42c356e6e763275432422dacbe

On CPU:
```
reward_to_go_jax 0.028926010942086577
reward_to_go_scipy 0.001616921043023467
scipy speedup 17.889563047554923
```

In the CPU case, for large values of N, the performance gets much closer to scipy lfilter.

On GPU:
```
reward_to_go_jax 5.656240948010236
reward_to_go_scipy 0.001609566854313016
scipy speedup 3514.1385602304745
```

`scipy.signal.lfilter` may not be a useful benchmark to compare against, but the jax version seems to get significantly slower on GPU.

This is using `git+https://github.com/google/jax.git@11c512a194add25ade20dbfacf635d6e0834eba3` and `https://storage.googleapis.com/jax-wheels/cuda92/jaxlib-0.1.15-cp37-none-linux_x86_64.whl`

**Comment by mattjj** [comment]
@jlebar Any gut reactions here?

I've been meaning to dump this JAX code (both in the OP and in @cshesse's gist) into HLO text for easier repros, but I haven't gotten to it yet. Let me know if you can take a look without that step :)

Wild guess: could these slowdowns be related to the TupleThunk issue you just fixed?

**Comment by jlebar** [comment]
It looks like this benchmark is including compilation time?

> Wild guess: could these slowdowns be related to the TupleThunk issue you just fixed?

Meaning, the TupleThunk change makes things slower?  That would be surprising.  Possible, I suppose, but I don't immediately see how.

Other than the benchmarking methodology question I don't have an immediate guess here.  I don't know enough about the difference in HLO implementation of the two functions to have any ideas, without actually looking at the HLO.

I'm always trying to think bigger picture, though. We're getting a fair number of these performance bugs, which is genuinely great.  And the bugs are also fairly high quality.  But how do we streamline the process of receiving them, triaging them, and acting upon them, for you and for us?

I'm wondering if e.g. we could have a script/document that people can follow when they're reporting a performance bug, which will capture the textual HLO and xla_hlo_profile?  Ideally it could capture some sort of execution trace too, but I'm not sure how much of that is OSS at the moment.

Thinking even bigger, if JAX had some sort of benchmark framework then it could take care of ensuring that we ignore the compilation time, in addition to doing all of the things above for bug reports.

**Comment by mattjj** [comment]
> Meaning, the TupleThunk change makes things slower? That would be surprising. Possible, I suppose, but I don't immediately see how.

Oh, sorry, I meant the opposite but my wording was bad: these numbers are all from before the TupleThunk fix, so I was guessing that your TupleThunk fix could have improved things.

Great thoughts on the big picture. We should at least have a doc, and I bet we can make a script, including maybe an internal one that collects a trace (which we can run ourselves on small repros). Maybe we (on the JAX side) can take this opportunity to start developing those things!

Thanks for weighing in, even with very little info from our side here.

**Comment by cshesse** [comment]
In my script, I call the function once during setup, is that not enough to avoid capturing compilation time?

For a benchmarking script, maybe a `jax.experimental.benchmark(fn)` vaguely like go bench setup:

```
import time
import jaxlib

first_run = True


def benchmark(fn, duration_sec=10):
    global first_run
    if first_run:
        print("jax version", jax.__version__)
        print("jaxlib version", jaxlib.__version__)
    first_run = False
    fn(1)
    n = 1
    while True:
        start = time.time()
        fn(n)
        elapsed = time.time() - start
        if elapsed > duration_sec:
            break
        n *= 10
    print("%s %d calls  %f us/call" % (fn.__name__, n, elapsed / n * 1e6))


def benchmark_reward_to_go_scipy(n):
    for _ in range(n):
        reward_to_go_scipy(rews, 0.99)


def benchmark_reward_to_go_jax(n):
    for _ in range(n):
        reward_to_go_jax(jrews, 0.99)


benchmark(benchmark_reward_to_go_scipy)
benchmark(benchmark_reward_to_go_jax)
```

**Comment by jlebar** [comment]
> In my script, I call the function once during setup, is that not enough
to avoid capturing compilation time?

Should be, yes.

**Comment by hawkinsp** [comment]
The main problem with the example in the first comment is that there is almost no useful work in the loop (a elementwise multiply and an add of `f32[16]` vector). You won't get good GPU utilization from this no matter what we do; CPU will be faster. The loop benchmark (using `scan`) is mostly measuring loop overheads; we could probably lower them but I am wondering if this code is actually representative of something you want to do. Put another way: even if the `scan` variant is 6x slower, does it actually matter?

**Comment by gehring** [comment]
@hawkinsp I've ran into this with bigger workloads. Specifically, a scan implementation lead to about a ~4x slow down, ~6x larger GPU memory usage, ~4-6x higher transfer rates as reported by `nvidia-smi` than an unrolled implementation. I'm not so certain that this is just an issue of higher overhead.

For a bit of context, the workload at each iterations was:
1.`extract_image_patch`
2. `tensordot(Nx28x28x9, Nx28x28x9x10 -> Nx28x28x10)`
3. `max(axis=-1)`
with batch size `N=10`. While the scan version compiles very quickly, the slow down makes it completely impractical. This was on an Nvidia 2080 Ti.

**Comment by gehring** [comment]
Additional observations: for the workload I describe earlier, if I run 25 unrolled steps per scan iteration, I still see a drastic increase in GPU usage, memory usage and CPU-GPU communication. Everything I've observed seems to hint at some memory reuse issue or some arrays placed on the wrong device issue. This issue is a big blocker for me. If there is anything I can do to help fix it, please let me know.

Note that while the reported GPU usage increases (from ~60% to 99%) when going from unrolled to scan, the actual performance is much much worse.

**Comment by jekbradbury** [comment]
@gehring Would it be possible to get us a self-contained example that demonstrates your slowdown? There are a few known drawbacks of XLA:GPU loops, but `scan` is typically supposed to be able to avoid the worst of it (raw `while_loop` loops have to bring their condition boolean back to the host at each step, while `scan` always has a fixed trip count).

**Comment by gehring** [comment]
@jekbradbury I'll try to strap a standalone example by the end of the week. In the meantime, if it makes any difference, I realize I forgot to mention that the workload I described is part of a SGD (pure python) loop, doing a forward and backwards pass in each jitted call.

**Comment by gehring** [comment]
@jekbradbury It ended up being much easier to do than I thought. The code where I first encountered this is somewhat complex but I was able to reproduce it with just a scan looping `amat @ x + bvec`. I saved it in a [notebook](https://colab.research.google.com/gist/gehring/bee2f5bd37c108c7620404e2c47652ed/jax-scan-issue.ipynb) for convenience.

On Google's public colab (GPU runtime), I get the following results:
```
Time taken unrolled: 12.955157 sec
Time taken scan: 197.164844 sec
```
I tried this example on a different workstation (with an old Quadro K2200) and observed pretty much the same symptoms as described earlier (using `nvidia-smi`).

**Comment by jekbradbury** [comment]
Thanks! This looks like a performance issue with JVP and possibly transpose of scan, which emit `dynamic-slice` and `dynamic-update-slice` operations (respectively) that appear to be the bottleneck here (and seemingly not necessary, since they don't show up in the unrolled version). Going to look into it, but I don't know this part of the code very well.

**Comment by gehring** [comment]
@jekbradbury Any progress or updates on this? No pressure, I'm just trying to get an idea as to whether a fix is in the works or if I should commit to a workaround.

**Comment by jekbradbury** [comment]
I had a partial solution in the linked PR, but we ended up deciding against that approach and sketched out a more direct (and complete) solution two weeks ago. I haven’t implemented it yet, though it’s been on my todos; I expect I’ll have a new PR sometime in the next week or so.
